### PR TITLE
Add External component

### LINKED ISSUE #469: `External` element of the storage.
At the moment to call a external contract, developer needs to store an `Address` and then construct `ContractRef` and only then can use it. I propose to simplify this flow, as it is very common.

```rust
#[odra::module]
pub struct TokenProxy {
    token: External<Cep78ContractRef>
}

#[odra::module]
impl TokenProxy {
    pub fn init(&mut self, token_address: Address) {
        self.token.set(token_address);
    }

    pub fn balance_of(&self, user: Address) -> U256 {
        self.token.balance_of(user)
    }

    pub fn get_token_address(&self) -> Option<Address> {
        self.token.get()
    }
}
```

When `self.token.set(token_address)` executes, then it should store only the `Address` to the storage, so functionally it behaves as `Var<Address>`.

### PR DESCRIPTION
**Pull request by kpob**
* Use External in modules and examples
* update macros
* add execution errors related to External

## Summary by CodeRabbit

- **New Features**
  - Introduced a new `External` module for handling and manipulating references to external contracts.
  - Enhanced `BalanceChecker` to include a `token` field and streamlined the `check_balance` method.

- **Bug Fixes**
  - Added new error variants `MissingAddress` and `AddressAlreadySet` to improve address handling.

- **Enhancements**
  - Updated `MockCep78Operator` to use the `External` trait, impacting method calls like `mint`, `transfer`, and `approve`.
  - Improved schema handling by adding `SchemaErrorsItem` and `SchemaEventsItem` for better error and event management.

- **Refactor**
  - Reorganized export declarations to include the new `External` entity and reordered others for clarity.

### CODE REVIEW

**Comment by kubaplas** [comment]
What if the external contract has a set() method?

**Review by kubaplas — APPROVED** [review]

File: examples/src/contracts/balance_checker.rs [hunk 1]
```diff
@@ -1,16 +1,24 @@
...
-pub struct BalanceChecker;
+pub struct BalanceChecker {
+    token: External<TokenContractRef>
+}
 
 #[odra::module]
 impl BalanceChecker {
+    /// Initializes the contract with the given parameters.
+    pub fn init(&mut self, address: Address) {
+        self.token.set(address);
```
Comment (kubaplas) [inline examples/src/contracts/balance_checker.rs:15]: What if ecternal contract has the `set()` method?

Comment (kpob) [reply examples/src/contracts/balance_checker.rs:15]: If a contract implements the `set()` method, you can't use the deref [coercion mechanism](https://doc.rust-lang.org/std/ops/trait.Deref.html#deref-coercion), so you call the `External.set()` method. In this case you need to dereference explicitly:  
```rust
use core::ops::DerefMut;
use odra::{Address, External, prelude::*};

#[odra::module]
pub struct Contract {
    ext: odra::External<SetterContractRef>
}

#[odra::module]
impl Contract {
    pub fn init(&mut self, address: odra::Address) {
        self.ext.set(address);
    }

    pub fn set(&mut self, value: bool) {
        self.ext.deref_mut().set(false);
        // or
        DerefMut::deref_mut(&mut self.ext).set(false);
    }
}

#[odra::external_contract]
pub trait Setter {
    fn set(&mut self, value: bool);
}
```

Comment (zie1ony) [reply examples/src/contracts/balance_checker.rs:15]: Please add this to the docs of `Extern`

**Review by zie1ony — APPROVED** [review]
Well done :)

**Pull request merged**